### PR TITLE
Fix broken setup_with_port

### PR DESCRIPTION
**Pull request by StoneDot**
*Description of changes:*
Previous `setup_with_port` actually fail to launch dynamodb-local.
This PR fixes broken `setup_with_port`. 

The main fix points are:

* `docker ps` was not able to filter intended container when old version docker is used. See [this PR](https://github.com/moby/moby/pull/40442).
* Variable `docker` was reused resulted in failure of `docker run`.
* We had to execute `cargo test` twice to pass tests.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tests/integration_test.rs [hunk 1]
```diff
@@ -19,54 +19,106 @@ use predicates::prelude::*; // Used for writing assertions
...
+        "--format",
+        "{{.Ports}}",
         "--filter",
         "ancestor=amazon/dynamodb-local",
     ]);
-    let check_out = check_cmd.output().expect("failed to execut check cmd");
-    if check_out.stdout.len() != 0 {
-        println!("DynamoDB Local is already running.");
+    let check_out = check_cmd.output().expect("failed to execute check cmd");
+    let reg_str = format!(r"(?m)^[.\d]+:{}->\d+/tcp$", port);
```
Comment (itazur) [inline tests/integration_test.rs:56]: In the IPv6 available environment, this regex pattern doesn't match correctly.

Only IPv4 available environment:
```
$ docker ps --format {{.Ports}} --filter ancestor=amazon/dynamodb-local
0.0.0.0:8000->8000/tcp
```

IPv4 and IPv6 available environment:
```
$ docker ps --format {{.Ports}} --filter ancestor=amazon/dynamodb-local
0.0.0.0:8000->8000/tcp, :::8000->8000/tcp
```

**Review by itazur — CHANGES_REQUESTED** [review]
Could you make the regex support the IPv6 available environment?
I'll take a look at the remaining part after that.

File: tests/integration_test.rs (hunk 1 above)
Comment (StoneDot) [reply tests/integration_test.rs]: I fixed the regex and test on Ubuntu 21.10.

**Comment by StoneDot** [comment]
Thank you for your feedback. I fixed the pointed out part.

**Comment by StoneDot** [comment]
I rebased and squash commits to resolve conflict.

File: tests/integration_test.rs [hunk 2]
```diff
@@ -19,54 +19,106 @@ use predicates::prelude::*; // Used for writing assertions
...
+    } else {
+        false
+    }
+}
+
+async fn setup_with_port(port: i32) -> Result<Command, Box<dyn std::error::Error>> {
+    // Check the current process at first to allow multiple threads to run tests concurrently
+    if check_dynamodb_local_running(port as u16).await {
         return Ok(Command::cargo_bin("dy")?);
     };
```
Comment (itazur) [inline tests/integration_test.rs:72]: We don't need to call `check_dynamodb_local_running()` twice, I think.

Assume that the first `check_dynamodb_local_running()` still exists here.
All multi-threaded tests run `check_dynamodb_local_running()` once for each at first.
When the docker container has been already created, they can return soon. That's good!
But when the docker container hasn't been created yet, the first test locks the mutex and creates a new docker container.
All the others need to wait it and then run `check_dynamodb_local_running()` again.

Then, assume that the first `check_dynamodb_local_running()` is removed.
Multi-threaded tests try to lock the mutex first, the only first one can lock and the others need to wait.
At first glance, this seems to lead to slow down.
But, when the docker container has already been created, the first one runs `check_dynamodb_local_running()` once and the mutex is **unlocked soon**, and then the others restart and run `check_dynamodb_local_running()` only once for each.
When the docker container hasn't been created yet, the first one runs `check_dymanodb_local_running()` once and creates a new docker container, and then the others run`check_dymanodb_local_running()` once for each.

Please let me know if you have any opinions and discussions.

**Review by itazur — CHANGES_REQUESTED** [review]
I've commented about the following points:
- `check_dynamodb_local_running()` doesn't need to be `async`.
- We don't need to call `check_dynamodb_local_running()` twice.

**Comment by StoneDot** [comment]
Double invocation of `check_dynamodb_local_running` is an intended implementation for performance reasons. I performed performance comparison like following.

`current` the means following code.
```rust
    // Check the current process at first to allow multiple threads to run tests concurrently
    if check_dynamodb_local_running(port as u16).await {
        return Ok(Command::cargo_bin("dy")?);
    };
```

`comment out` means  the following code.
```rust
    // Check the current process at first to allow multiple threads to run tests concurrently
    // if check_dynamodb_local_running(port as u16).await {
    //     return Ok(Command::cargo_bin("dy")?);
    // };
```

## Performance comparison
### Performance comparison on Mac
`current`
```
1st
test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.48s
cargo test  1.83s user 1.34s system 179% cpu 1.764 total

2nd
test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.46s
cargo test  1.77s user 1.29s system 173% cpu 1.759 total

3rd
test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.45s
cargo test  1.76s user 1.01s system 172% cpu 1.605 total
```

`comment out`
```
1st
test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 3.37s
cargo test  1.74s user 0.87s system 73% cpu 3.540 total

2nd
test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 3.38s
cargo test  1.74s user 0.85s system 73% cpu 3.547 total

3rd
test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 3.60s
cargo test  1.75s user 0.85s system 69% cpu 3.762 total
```

Average finished time (current): 1.463s
Average finished time (comment out): 3.450s
Average total time (current): 1.709s
Average total time (comment out): 3.616s

### Performance comparison on Windows
Added later

File: tests/integration_test.rs [hunk 3]
```diff
@@ -19,54 +19,106 @@ use predicates::prelude::*; // Used for writing assertions
...
 
-fn setup_with_port(port: i32) -> Result<Command, Box<dyn std::error::Error>> {
-    // NOTE: setup() spins up DynamoDB Local in 8000 port as "local" region assumes it. Better to make it configurable.
-    let mut docker = Command::new("docker");
+// We use std::sync::Mutex instead of tokio::sync::Mutex, because mutex must be poisoned after setup failure.
+static SETUP_MUTEX: Lazy<Mutex<i32>> = Lazy::new(|| Mutex::new(0));
 
-    let check_cmd = docker.args(&[
+/// Check existence of docker process for dynamodb-local
+async fn check_dynamodb_local_running(port: u16) -> bool {
```
[the remark this answers is not in the document]
Comment (StoneDot) [reply tests/integration_test.rs:45]: Thank you for your point out. I will fix this.

File: tests/integration_test.rs (hunk 2 above)
Comment (StoneDot) [reply tests/integration_test.rs:72]: As I describe in https://github.com/awslabs/dynein/pull/28#issuecomment-972880324, this is a intended implementation for performance reasons. Although, I will add a comment for later reference.
Please let me know if you have any other question regarding this implementation.

**Comment by StoneDot** [comment]
I rebased again to resolve the conflict.

**Comment by itazur** [comment]
LGTM. Thanks!

**Review by itazur — COMMENTED** [review]
LGTM. Thanks!

**Review by itazur — COMMENTED** [review]
LGTM. Thanks!

**Pull request merged**